odl/tomo/data/uncompr_bin.py

```diff
@@ -0,0 +1,472 @@
+
+        # Get byte range and set start
+        byte_range = field[keys['byte_range']].split('-')
+        byte_start = int(byte_range[0]) - 1
+        byte_end = int(byte_range[-1]) - 1
+        new_field['offset_bytes'] = byte_start
+
+        # Data type: transform to Numpy format and get shape from its
+        # itemsize and the byte range
+        dtype = dtype_map[field[keys['dtype']]]
+        byte_size = byte_end - byte_start + 1
+
+        if hasattr(dtype, 'itemsize') and byte_size % dtype.itemsize:
+            raise ValueError(
+                'in row {}: byte range {} not a multiple of itemsize {} '
+                'of the data type {}.'
+                ''.format(row + 1, field[keys['byte_range']], dtype.itemsize,
+                          field[keys['dtype']]))
+
+        new_field['dtype'] = dtype
+        new_field['size_bytes'] = byte_size
+        # Assuming 1d arrangement of multiple elements
+        # TODO: find way to handle 2d fields
+        if nend:
+            new_field['dshape'] = (nend - nstart + 1,)
+        elif hasattr(dtype, 'itemsize'):
+            new_field['dshape'] = (byte_size / dtype.itemsize,)
+        else:
+            new_field['dshape'] = (1,)
+
+        conv_list.append(new_field)
+
+    return tuple(conv_list)
+
+
+class FileReaderUncompressedBinary(object):
+
+    """Reader for uncompressed binary files including a header."""
+
+    def __init__(self, file, header_fields):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        file : file-like or str
+            Stream or filename from which to read the data. The stream
+            is allowed to be already opened in 'rb' mode.
+        header_fields : sequence of dicts
+            Definition of the fields in the header (per row), each
+            containing key-value pairs for the following keys:
+
+            - ``'name'`` : Label for the field.
+            - ``'offset_bytes'`` : Start of the field in bytes.
+            - ``'size_bytes'`` : Size of the field in bytes.
+            - ``'dtype'`` : Data type in Numpy- or Numpy-readable format.
+            - ``'dshape'`` (optional) : The array of values is reshaped to
+              this shape.
+            - ``'description'`` : A human-readable description of the field.
+        """
+        try:
+            f = open(file, 'rb')
+        except TypeError:
+            f = file
+
+        if f.mode != 'rb':
+            raise ValueError("`file` must be opened in 'rb' mode, but mode "
+                             "'is {}'".format(f.mode))
+
+        self.file = f
+
+        # Initialize some attributes to default values, plus the header
+        self.data_shape = -1  # Makes reshape a no-op
+        self.header_bytes = None
+        self.header_fields = header_fields
+        self.data = None
+        if self.header_fields:
+            self.header = self.read_header()
+        else:
+            self.header = None
+
+    @classmethod
```

Comment:
I'll see if I can use the standard initializer for this case, too. 

Generally class methods `from...` are the standard way of having alternative constructors for classes, see e.g. [dict.fromkeys](https://docs.python.org/3.5/library/stdtypes.html#dict.fromkeys). Those alternative constructors are used whenever the construction pattern (mostly input parameters) contradict the one of the standard constructor, which may either be inherited or designed for subclassing (like in this case). In fancy words, alternative constructors are needed whenever they would violate the [Liskov substitution principle](https://en.wikipedia.org/wiki/Liskov_substitution_principle) - good practice is to compartmentalize these violators.

That said, I'll see if we really need it.